src/shadowbox/server/manager_metrics.ts

```diff
@@ -40,4 +72,74 @@ export class PrometheusManagerMetrics implements ManagerMetrics {
     }
     return {bytesTransferredByUserId: usage};
   }
+
+  async getServerMetrics({hours}: ServerMetricsTimeframe): Promise<ServerMetrics> {
+    const dataTransferredByLocation = await this.prometheusClient.query(
+      `sum(increase(shadowsocks_data_bytes_per_location{dir=~"c<p|p>t"}[${hours}h])) by (location, asn, asorg)`
+    );
+    const tunnelTimeByLocation = await this.prometheusClient.query(
+      `sum(increase(shadowsocks_tunnel_time_seconds_per_location[${hours}h])) by (location, asn, asorg)`
+    );
+    const dataTransferredByAccessKey = await this.prometheusClient.query(
+      `sum(increase(shadowsocks_data_bytes{dir=~"c<p|p>t"}[${hours}h])) by (access_key)`
+    );
+    const tunnelTimeByAccessKey = await this.prometheusClient.query(
+      `sum(increase(shadowsocks_tunnel_time_seconds[${hours}h])) by (access_key)`
+    );
+
+    const server = [];
+    for (const entry of dataTransferredByLocation.result) {
+      const result: ServerMetricsServerEntry = {
+        location: entry.metric['location'],
+        asn: parseInt(entry.metric['asn']),
+        asOrg: entry.metric['asorg'],
+        dataTransferred: {
+          bytes: parseFloat(entry.value[1]),
+        },
+      };
+
+      const matchingTunnelTimeResult = tunnelTimeByLocation.result.find((target) => {
```

Comment:
You are doing O(n * m). This will collapse with servers with 10s of thousands of keys.
Please change it so it's O(n + m). Create a map and run an loop for each metric

Same for the access key metrics.